end-to-end-tests/src/helpers/ctx.rs

```diff
@@ -84,32 +86,168 @@ pub fn nexus_addr() -> SocketAddr {
+    let silo_name = config.recovery_silo.silo_name.as_str();
+    let username: oxide_client::types::UserId =
+        config.recovery_silo.user_name.as_str().parse().map_err(|s| {
+            anyhow!("parsing configured recovery user name: {:?}", s)
+        })?;
+    // See the comment in the config file.
+    let password: oxide_client::types::Password = "oxide".parse().unwrap();
+    let login_request_body =
+        serde_json::to_string(&UsernamePasswordCredentials {
+            username: username,
+            password: password,
+        })
+        .context("serializing login request body")?;
+
+    // Do not have reqwest follow redirects.  That's because our login response
+    // includes both a redirect and the session cookie header.  If reqwest
+    // follows the redirect, we won't have a chance to get the cookie.
+    let reqwest_login_client = reqwest::ClientBuilder::new()
+        .connect_timeout(Duration::from_secs(15))
+        .redirect(reqwest::redirect::Policy::none())
+        .timeout(Duration::from_secs(60))
+        .build()?;
+    let login_url = format!("{}/login/{}/local", base_url, silo_name);
+
+    // By the time we get here, we generally would have successfully resolved
+    // Nexus's external IP address from the external DNS server.  So we'd
+    // expect Nexus to be up.  But that's not necessarily true: external DNS can
+    // be set up during rack initialization, before Nexus has opened its
+    // external listening socket.  This is arguably a bug, advertising a service
+    // before it's ready, but a pretty niche corner case (rack initialization)
+    // and anyway DNS is always best-effort.  The point is: let's retry a little
+    // while if we can't immediately connect.
+    let response = wait_for_condition(
+        || async {
+            // Use a raw reqwest client because it's not clear that Progenitor
+            // is intended to support endpoints that return 300-level response
+            // codes.  See progenitor#451.
+            reqwest_login_client
+                .post(&login_url)
+                .body(login_request_body.clone())
+                .send()
+                .await
+                .map_err(|e| {
+                    if e.is_connect() {
+                        CondCheckError::NotYet
+                    } else {
+                        CondCheckError::Failed(
+                            anyhow::Error::new(e).context("logging in"),
+                        )
+                    }
+                })
+        },
+        &Duration::from_secs(1),
+        &Duration::from_secs(30),
+    )
+    .await
+    .context("logging in")?;
+
+    let session_cookie = response
+        .headers()
+        .get(http::header::SET_COOKIE)
+        .ok_or_else(|| anyhow!("expected session cookie after login"))?
+        .to_str()
+        .context("expected session cookie token to be a string")?;
+    let (session_token, rest) = session_cookie.split_once("; ").unwrap();
```

Comment:
Can be a follow up but would be nice if `oxide_client` wrapped some of this up.

Reply:
Yeah, fair question.  At first I thought you meant the progenitor client, where the plan is sort of uncertain (see oxidecomputer/progenitor#451), but I think you meant the oxide_client workspace crate, where we could definitely put something like this.

I'm not clear how reasonable an idea it is to use the login endpoint in this way (and that's part of the uncertainty around having progenitor support it).  In a real system we really expect people to use the console to log in and use a device authn token for programmatic access.  I think it's just automated rack setup where we can't do that.  So I think I'll hold off for now.